Add Writer.embed function to allow writing pre-formed XML without escaping anything

MFAshby opened the pull request:
Useful for embedding one XML fragment into another

MFAshby committed:
Add Writer.embed function to allow writing pre-formed XML without
escaping anything

Useful for embedding one XML fragment into another

ianprime0509 commented on src/Writer.zig:
```diff
@@ -183,6 +183,17 @@ pub fn text(writer: *Writer, s: []const u8) anyerror!void {
 
+// insert some existing XML document without escaping anything
+pub fn embed(writer: *Writer, s: []const u8) anyerror!void {
+    switch (writer.state) {
+        .after_structure_end, .text => {},
+        .element_start => try writer.raw(">"),
+        .start, .after_bom, .after_xml_declaration, .end => unreachable,
+    }
```
```suggestion
    switch (writer.state) {
        .start, .after_bom, .after_xml_declaration, .after_structure_end, .text, .end => {},
        .element_start => try writer.raw(">"),
    }
```
I don't see any reason to restrict this function to operate only inside the root element: it's reasonable that someone might want to use it to embed a DTD at the beginning of the document, for example, which isn't currently possible with the other methods.

MFAshby replied:
That's a good point, I was thinking too narrowly about my own use case, embedding xml fragments inside an existing document. Fixed in https://github.com/ianprime0509/zig-xml/pull/39/commits/f54c606ce886ac26d6dc0e902733c1aa71a5013e

ianprime0509 commented on src/Writer.zig:
```diff
@@ -183,6 +183,17 @@ pub fn text(writer: *Writer, s: []const u8) anyerror!void {
+        .start, .after_bom, .after_xml_declaration, .end => unreachable,
+    }
+    try writer.raw(s);
+    writer.state = .after_structure_end;
```
```suggestion
    writer.state = switch (writer.state) {
        .start, .after_bom, .after_xml_declaration => .after_xml_declaration,
        .element_start, .after_structure_end, .text => .text,
        .end => .end,
    };
```
Once this function is opened up to work in more scenarios, other possible states also need to be handled. In this suggestion, I chose `text` as the ending state within the root element to avoid indentation if another element is started right after this function (to maximize the possible uses, including for embedding raw text nodes).

ianprime0509 reviewed (changes requested):
Thanks for the PR! I think this is a good function to have. I have a couple suggestions to expand the possible use-cases for it. Also, it should be added to `GenericWriter` so users of that API can call it: https://github.com/ianprime0509/zig-xml/blob/ee8e09bccf73fa2b494e7e4a706a5ca5523f8eac/src/xml.zig#L433-L471

MFAshby commented:
@ianprime0509 thanks for the review, I'm glad this can be useful. I'll apply your suggestions shortly!

MFAshby committed:
Add missing "embed" function to GenericWriter

Fixes: https://github.com/ianprime0509/zig-xml/pull/39#pullrequestreview-2397316258

Make the "embed" function more permissive about contexts in which it's
used, as this makes it more flexible

Fixes:
- https://github.com/ianprime0509/zig-xml/pull/39#discussion_r1817901534
- https://github.com/ianprime0509/zig-xml/pull/39#discussion_r1817902115

MFAshby committed:
Add a simple test for "embed" function in Writer

Add a 'testbed' facility for writing more test.

MFAshby commented:
Thanks again @ianprime0509 I've added your suggestions in https://github.com/ianprime0509/zig-xml/pull/39/commits/f54c606ce886ac26d6dc0e902733c1aa71a5013e

I've also started adding tests for Writer as there didn't seem to be any https://github.com/ianprime0509/zig-xml/pull/39/commits/1a0355a7f3137b087b4f6d4c2e06d14c6e9f07c6, but happy to drop this commit or move it to a new PR if it's not something you're interested in right now.

ianprime0509 commented:
Awesome, thanks! Yeah, I haven't added any tests or proper doc comments to the writer yet, unfortunately. I'll probably go a slightly different direction with the tests when I get to it (focusing on having one self-contained doctest for each function, similar to the current reader tests), but I think what you have for now is a useful start.
